### PR TITLE
TypeScript reorg, without the kernel abstraction

### PR DESCRIPTION
**Pull request by ElvishJerricco**

### CODE REVIEW

**Comment by ElvishJerricco** [comment]
@angerman, @dfordivam, do you think we should be checking in the `package-lock.json` file? It's basically like a cabal freeze file for NPM.

File: src/node_runner.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { configureFileSystem, Process } from "./index";
+import * as process from "process";
+import * as fs_ from "fs";
+
+async function main(): Promise<void> {
+  try {
+    var args = process.argv.splice(process.execArgv.length + 2);
+
+    if (args.length == 0) {
+      console.log('Usage: node run_node.js <wasm-executable-name>');
```
Comment (dfordivam) [inline src/node_runner.ts]: node_runner.js

File: src/DeviceFileSystem.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import { BaseFileSystem, BFSCallback, FileSystem, FileSystemOptions
+       } from "browserfs/dist/node/core/file_system";
+import Stats from 'browserfs/dist/node/core/node_fs_stats';
+import { File } from "browserfs/dist/node/core/file";
+import { FileFlag } from "browserfs/dist/node/core/file_flag";
+import { ApiError } from 'browserfs/dist/node/core/api_error';
+
+export interface Device {
```
Comment (dfordivam) [inline src/DeviceFileSystem.ts:8]: So Device is a completely new concept, it is not in BrowserFS. I see that it has a 'File'

I think we will have to implement a 'File' interface, lets say 'JsaddleFile'.
So why have 'Device', I think doing 'files: {[name: string]: File};' would be sufficient?

**Review by dfordivam — COMMENTED** [review]
Not quite sure why we need the 'Device' interface.

**Comment by ElvishJerricco** [comment]
@dfordivam The difference between `Device` and `File` is 1) you can `stat` a device, providing metadata, and 2) a driver can hold off on initializing anything until the file is `open`ed. Plus, I'm not sure if there would be complications with trying to reuse the same `File` instance if it gets closed and reopened.

**Comment by ElvishJerricco** [comment]
@dfordivam Also note that `interface`s in TypeScript are more lightweight than in other languages. You can generate them ad-hoc.

```typescript
let myDevice: Device = {
  open: flag => { ... },
  stat: isLstat => { ... }
}
```

File: src/DeviceFileSystem.ts (hunk 2 above)
Comment (dfordivam) [reply src/DeviceFileSystem.ts:8]: something like this https://github.com/jvilk/BrowserFS/blob/28a26cbcb89b5cef59a586f55045aabe98561f47/src/backend/Emscripten.ts#L34
https://github.com/jvilk/BrowserFS/blob/28a26cbcb89b5cef59a586f55045aabe98561f47/src/backend/Emscripten.ts#L253

**Comment by dfordivam** [comment]
> @dfordivam The difference between `Device` and `File` is 1) you can `stat` a device, providing metadata, and 2) a driver can hold off on initializing anything until the file is `open`ed. Plus, I'm not sure if there would be complications with trying to reuse the same `File` instance if it gets closed and reopened.

1) I see that in this implementation the stat on File ends up getting info from FS (https://github.com/jvilk/BrowserFS/blob/28a26cbcb89b5cef59a586f55045aabe98561f47/src/backend/Emscripten.ts#L71)
   so stat can be maintained inside FS.
   (Edit: But since in our case the `stat` will always return some constant object, it can be simplified and stored just at a `Device` object level. In that case we will have to fetch the data from `_device` while implementing `stat` for `JSaddleFile`. This will avoid keeping a reference of filesystem `_fs` at the File level.)

2) I think this again is an implementation detail at FileSystem level. When we initialize the FileSystem we are providing the 'files' object. 
  The 'close' of a file will be just removing a FileDescriptor from map. So if I open, close and open a file again, ideally the  filesystem code should be smart enough to not modify anything in the 'files' object or the actual 'JSaddleFile' objects.

**Comment by ElvishJerricco** [comment]
@dfordivam wouldn't we have to implement a new FileSystem implementation for every device we wanted that way?

**Comment by dfordivam** [comment]
@ElvishJerricco 
The `files` object in `FileSystem` can store 'devices' of different kinds all implementing the `File` interface. And as far I understand a new 'device' implementing `File` can implement the interface APIs like `open`, `stat`, `close` as it wants.

**Comment by ElvishJerricco** [comment]
@dfordivam But notice that the `File` interface doesn't have any kind of `open` method. Once the device is closed, it can't be reopened that way. Or at least, there's no way to tell the device to re-initialize itself.

**Comment by dfordivam** [comment]
@ElvishJerricco ok, this is a fair use case. 
The other way of solving this with `export class DeviceFile extends BaseFile implements File` doesn't look that appealing, so having an additional layer of `Device` is good for now.

**Pull request merged**